webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala

```diff
@@ -541,41 +566,45 @@ object ConstructResponseUtilV2 {
         }
 
         /**
-          * Given a resource IRI, finds any link values in the resource, and recursively embeds the target resource in each link value.
-          *
-          * @param resourceIri      the IRI of the resource to start with.
-          * @param alreadyTraversed a set (initially empty) of the IRIs of resources that this function has already
-          *                         traversed, to prevent an infinite loop if a cycle is encountered.
-          * @return the same resource, with any nested resources attached to it.
-          */
+         * Given a resource IRI, finds any link values in the resource, and recursively embeds the target resource in each link value.
+         *
+         * @param resourceIri      the IRI of the resource to start with.
+         * @param alreadyTraversed a set (initially empty) of the IRIs of resources that this function has already
+         *                         traversed, to prevent an infinite loop if a cycle is encountered.
+         * @return the same resource, with any nested resources attached to it.
+         */
         def nestResources(resourceIri: IRI, alreadyTraversed: Set[IRI] = Set.empty[IRI]): ResourceWithValueRdfData = {
-            val resource = flatResourcesWithValues(resourceIri)
+            val resource = visibleResources(resourceIri)
 
             val transformedValuePropertyAssertions: RdfPropertyValues = resource.valuePropertyAssertions.map {
                 case (propIri, values) =>
-                    val transformedValues = values.map {
-                        value =>
+                    val transformedValues: Seq[ValueRdfData] = values.foldLeft(Vector.empty[ValueRdfData]) {
```

Comment:
The whole `splitMainResourcesAndValueRdfData` method has always been a headache for me to read: too long and too much nesting. Each time I have to work on it, I try to make it a bit clearer, but it really needs to be split up into much smaller pieces. I was going to ask you to refactor it with Tobias, since he's the author. :) For now, I've just moved the `nestResources` method out to the top level, and I've moved the `foldLeft` operation into another function, `transformValuesByNestingResources` (in 6e96460). But even after doing that, `splitMainResourcesAndValueRdfData` is still 273 lines long. :(

Reply:
> For now, I've just moved the `nestResources` method out to the top level, and I've moved the `foldLeft` operation into another function, `transformValuesByNestingResources` (in [6e96460](https://github.com/dasch-swiss/knora-api/commit/6e964606936157a6198f41a7250bb86f58c6065f)). But even after doing that, `splitMainResourcesAndValueRdfData` is still 273 lines long. :(

It took me really long time to understand what the `splitMainResourcesAndValueRdfData` does, but at least now with `transformValuesByNestingResources` it is a tiny bit easier to read.